app/src/main/java/com/eis/phoneringer/MainActivity.java

```diff
@@ -1,14 +1,126 @@
+ * @author Alberto Ursino
+ */
 public class MainActivity extends AppCompatActivity {
 
+    private Button ringButton;
+    private EditText phoneNumber;
+    private static final String DEFAULT_PASSWORD = "password";
+    private static final int WAIT_TIME_PERMISSION = 1500;
+    private PasswordManager passwordManager;
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
+
+        final Context context = getApplicationContext();
+        ringButton = findViewById(R.id.ring_button);
+        phoneNumber = findViewById(R.id.phone_number);
+        passwordManager = new PasswordManager(context);
+
+        //Some tricks with permissions
+        if (checkPermission()) {
+            SMSHandler.getInstance().setup(context);
+            SMSHandler.getInstance().setReceivedListener(new ReceivedMessageListener(context));
+        } else {
+            requestPermission();
+        }
+
+        //Default password equals for all the devices
+        if (passwordManager.getPassword().isEmpty())
+            passwordManager.setPassword(DEFAULT_PASSWORD);
+
+        ringButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                sendCommand();
+            }
+        });
+    }
+
+    /**
+     * Sends a ring command to the phone number given by the user
+     */
+    private void sendCommand() {
+        final Context context = getApplicationContext();
+        //Callback when the command is sent
+        SMSSentListener smsSentListener = new SMSSentListener() {
+            @Override
+            public void onSMSSent(SMSMessage message, SMSMessage.SentState sentState) {
+                Toast.makeText(context, getString(R.string.sent_listener_toast) + phoneNumber.getText().toString(), Toast.LENGTH_SHORT).show();
+            }
+        };
+        //Passing to the AppManager a new RingCommand built through the phone number given by the user
+        try {
+            AppManager.getInstance().sendCommand(context, new RingCommand(new SMSPeer(phoneNumber.getText().toString()), passwordManager.getPassword()), smsSentListener);
+        } catch (InvalidTelephoneNumberException e) {
+            Toast.makeText(context, getString(R.string.invalid_phone_number_toast), Toast.LENGTH_SHORT).show();
+        } catch (InvalidSMSMessageException e) {
+            Toast.makeText(context, getString(R.string.invalid_sms_message_toast), Toast.LENGTH_SHORT).show();
+        }
     }
+
+    /**
+     * @return true if the app has permissions, false otherwise
+     */
+    public boolean checkPermission() {
+        return !((getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED) ||
+                !(getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED));
```

Comment:
This is too complicated to read and I think it might be wrong, what you want to do is check if both equivalences are true. What you're doing is ¬(P∨¬Q), but it should be P∧Q. Actually, you're doing the same equivalence twice. This should just be
```
return getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED
```

Reply:
he probably wanted to type Manifest.permission.SEND_SMS. Anyway this is wrong